pkg/net/libp2p/libp2p_transport.go

```diff
@@ -219,38 +265,46 @@ func (t *Transport) openStream(ctx context.Context, p peer.ID) (network.Stream,
 	return nil, fmt.Errorf("failed to open stream to %s: %w", p, streamErr)
 }
 
-func (t *Transport) Send(dest types.NodeID, payload *messagepb.Message) error {
-	outBytes, err := proto.Marshal(payload)
-	if err != nil {
-		return fmt.Errorf("failed to marshal message: %w", err)
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, payload *messagepb.Message) error {
+	found := t.nodeExists(dest)
+	if !found {
+		return fmt.Errorf("failed to get address for node %s", dest)
 	}
```

Comment:
The difference is that if the target node is not found it is not possible to add this node within libp2p. And because of that, it is not possible to open a stream. Only the caller can fix it somehow (e.g., via Connect function).

Does it make sense?

>  handled by t.connectToNode.
That's true. But `connectToNode` doesn't return an error. And because of that and async nature of that function, we will run
```
go t.connectToNode(ctx, dest, t.connWg)
		return fmt.Errorf("no open stream for node %s, reopening", dest)
```
Then we get an error "that node doesn't exist" in logs and get "no open stream for node" error in the caller function.

As a result, I think we should distinguish between cases when the node doesn't exist at all and the case when the node exists but the stream doesn't

Reply:
Yes I agree that this potentially might be a useful thing to distinguish, but unless it happens atomically (checking both things while holding the lock), one can always construct a scenario where the first check is futile.
E.g., the check succeeds, releases the lock, and then someone removes the node by calling `CloseOldConnections`.

If we do want to distinguish it properly, I think the check still needs to happen atomically (for both the presence of the nodeInfo and the connection within). For example by extending the `t.getStream` function so we can make this distinction based on what it returns.

In fact, in any case, it still seems possible that the stream is closed *concurrently* by `CloseOldConnections` when `Send` is writing to it. It would be good to check with the doc that this doesn't panic and behaves in a predictable way.

Reply:
Agree. Fixed